File: src/Starter/Views/MainWindow.axaml.fs
namespace Starter.Views

open System
open Starter
open Starter.Controls
open Starter.Features
open Starter.Features.Logging
open Starter.Features.PlatformInterop

open System.Collections.Generic
open System.Windows.Input

open Avalonia
open Avalonia.Input
open Avalonia.Interactivity
open Avalonia.Controls
open Avalonia.Markup.Xaml
open Avalonia.VisualTree
open R3

type MainWindow() as this =
    inherit Window()

    static let normalResourceDictionary = ResourceDictionary()
    static let zoomedResourceDictionary = ResourceDictionary()
    static do
        normalResourceDictionary.Add("CornerRadius", CornerRadius 17)
        normalResourceDictionary.Add("SearchResultCornerRadius", CornerRadius 11)
        normalResourceDictionary.Add("ResultsPadding", Thickness(7))
        normalResourceDictionary.Add("SeparatorPadding", Thickness(13, 0))
        normalResourceDictionary.Add("SearchEnginePillMargin", Thickness(-3, 0, 10, 0))
        normalResourceDictionary.Add("TextBoxMargin", Thickness(0, 14, 13, 14))
        normalResourceDictionary.Add("TextBoxFontSize", 18.)
        normalResourceDictionary.Add("TextBoxLineHeight", 24.)
        normalResourceDictionary.Add("GridMargin", Thickness(15, 0, 0, 0))
        normalResourceDictionary.Add("SearchResultIconSize", 30.)
        normalResourceDictionary.Add("SearchResultPadding", Thickness(10, 9))
        normalResourceDictionary.Add("SearchResultFontSize", 13.)

        zoomedResourceDictionary.Add("CornerRadius", CornerRadius 17)
        zoomedResourceDictionary.Add("SearchResultCornerRadius", CornerRadius 10)
        zoomedResourceDictionary.Add("ResultsPadding", Thickness(8))
        zoomedResourceDictionary.Add("SeparatorPadding", Thickness(15, 0))
        zoomedResourceDictionary.Add("SearchEnginePillMargin", Thickness(-3, 0, 10, 0))
        zoomedResourceDictionary.Add("TextBoxMargin", Thickness(0, 14, 15, 15))
        zoomedResourceDictionary.Add("TextBoxFontSize", 20.)
        zoomedResourceDictionary.Add("TextBoxLineHeight", 27.)
        zoomedResourceDictionary.Add("GridMargin", Thickness(18, 0, 0, 0))
        zoomedResourceDictionary.Add("SearchResultIconSize", 35.)
        zoomedResourceDictionary.Add("SearchResultPadding", Thickness(10, 10))
        zoomedResourceDictionary.Add("SearchResultFontSize", 15.)

    do this.InitializeComponent()

    member this.ViewModel =
        match this.DataContext with
        | null -> failwith "No DataContext attached"
        | dc -> dc :?> ViewModels.MainWindowViewModel
    member this.TextBox = this.Get<TextBox> "TextBox"
    member this.SearchEnginePill = this.Get<SearchEnginePill> "SearchEnginePill"

    member this.ResultList = this.Get<ListBox> "ResultList"
    member this.ResultListScrollViewer() =
        this.ResultList.GetVisualDescendants()
        |> Seq.find (fun visual -> visual.Name = "PART_ScrollViewer")
        :?> ScrollViewer

    member private this.InitializeComponent() =
        AvaloniaXamlLoader.Load(this)
        #if DEBUG
        this.AttachDevTools()
        #endif

        match PlatformInteropFactory.GetPlatformInterop() with
        | :? Windows as platform -> platform.SetupHotkeyCallback this
        | :? Linux as platform -> platform.SetupHotkeyCallback this
        | _ -> ()

        this.Loaded.Add(fun _ ->
            this.SetupKeyboardShortcuts()

            // Bind config changes
            this.ViewModel.Config.Subscribe(fun config ->
                this.TransparencyLevelHint <-
                    match config.Background with
                    | Config.Background.Acrylic -> [| WindowTransparencyLevel.AcrylicBlur |].AsReadOnly()
                    | Config.Background.Mica -> [| WindowTransparencyLevel.Mica |].AsReadOnly()
                    | Config.Background.None -> [| WindowTransparencyLevel.None |].AsReadOnly()

                this.Resources <-
                    match config.ZoomedMode with
                    | false -> normalResourceDictionary
                    | true -> zoomedResourceDictionary
            )
            |> ignore

            // Bind single-search-engine pill
            this.ViewModel.CurrentActivator.Subscribe(fun singleSeMode ->
                match singleSeMode with
                | None ->
                    this.SearchEnginePill.IsVisible <- false
                    this.SearchEnginePill.DataContext <- null
                | Some se ->
                    this.SearchEnginePill.DataContext <- se
                    this.SearchEnginePill.IsVisible <- true
            )
            |> ignore

            // Subscribe to commands
            this.ViewModel.ClearTextBoxCommand
                .ObserveOnUIThreadDispatcher()
                .Subscribe(fun prefixLength ->
                    this.TextBox.Text <-
                        match this.TextBox.Text with
                        | null -> ""
                        | s -> s.Substring(prefixLength)
                    this.TextBox.CaretIndex <- this.TextBox.CaretIndex - prefixLength
                )
            |> ignore

            // Subscribe to pointer pressed events
            this.ResultList.AddHandler(
                InputElement.PointerReleasedEvent,
                EventHandler<PointerReleasedEventArgs>(fun s e ->
                    e.GetPosition(this)
                    |> this.GetVisualsAt
                    |> Seq.tryFind (fun v -> v.DataContext :? ViewModels.SearchResultViewModel)
                    |> Option.bind (_.DataContext >> Option.ofObj)
                    |> Option.iter (fun dataContext ->
                        dataContext
                        :?>  ViewModels.SearchResultViewModel
                        |> this.ViewModel.ValidateResult
                    )
                ),
                RoutingStrategies.Tunnel
            )
        )

        this.Activated.Add (fun _ ->
            // Center window
            match this.Screens.Primary with
            | null ->
                logger.Error "No screen available. Can't center window"
                failwith "No screen available. Can't center window"
            | screen ->
                this.Position <-
                    PixelPoint(
                        round ((float screen.WorkingArea.Width - (this.Width * screen.Scaling)) / 2.) |> int,
                        float screen.WorkingArea.Height * (5./16.) |> int
                    )

            // Reset focus
            this.TextBox.Focus() |> ignore
            this.TextBox.SelectAll()
            this.ResultList.Selection.Select 0 // Reset selection
            if this.ResultList.ItemCount <> 0 then
                this.ResultListScrollViewer().ScrollToHome() // Scroll to top to preserve the top padding
        )
        #if !DEBUG
        this.Deactivated.Add (fun _ -> this.Hide())
        #endif

    member private this.SetupKeyboardShortcuts() =
        // Subscribe to hide command
        this.ViewModel.HideCommand.Subscribe(fun _ ->
            Threading.Dispatcher.UIThread.Post(fun _ ->
                this.Hide()
            )
        )
        |> ignore

        // Add escape key binding
        let onEscape = // Run when "Escape" is pressed
            ReactiveUI.ReactiveCommand.Create(fun () ->
                if this.ViewModel.CurrentActivator.Value.IsSome then
                    this.ViewModel.ResetActivator()
                else
                    (this.ViewModel.HideCommand :> ICommand).Execute()
            )

        KeyBinding(
            Command = onEscape,
            Gesture = KeyGesture.Parse "Escape"
        )
        |> this.KeyBindings.Add

        // Set keyboard navigation
        // -> The goal is to be able to navigate in the listbox without losing the focus on the textbox
        let r = this.ResultList
        let d = System.EventHandler<KeyEventArgs>(fun _ e ->
            let newSelectedIdx =
                match e.PhysicalKey with
                | PhysicalKey.Backspace ->
                    if this.ViewModel.CurrentActivator.Value.IsSome && this.TextBox.CaretIndex = 0 then
                        this.ViewModel.ResetActivator()
                        e.Handled <- true

                    None

                | PhysicalKey.Tab when this.ViewModel.SearchResults.Count <> 0 ->
                    match e.KeyModifiers &&& KeyModifiers.Shift = KeyModifiers.Shift with
                    | true -> (r.SelectedIndex - 1) |> max 0 |> Some
                    | _ -> (r.SelectedIndex + 1) |> min (r.ItemCount - 1) |> Some

                | _ when this.ViewModel.SearchResults.Count <> 0 ->
                        match e.Key.ToNavigationDirection() |> Option.ofNullable with
                        | Some NavigationDirection.Up ->
                            (r.SelectedIndex - 1)
                            |> max 0
                            |> Some
                        | Some NavigationDirection.Down ->
                            (r.SelectedIndex + 1)
                            |> min (r.ItemCount - 1)
                            |> Some
                        | _ -> None

                | _ -> None

            match newSelectedIdx with
            | None -> ()
            | Some newSelectedIdx ->
                // Scroll to top or bottom to preserve the paddings
                let scrollViewer = this.ResultListScrollViewer()
                if newSelectedIdx = 0 then
                    scrollViewer.ScrollToHome()
                elif newSelectedIdx = r.ItemCount - 1 then
                    scrollViewer.ScrollToEnd()

                r.Selection.SelectedIndex <- newSelectedIdx
                e.Handled <- true
        )
        this.TextBox.AddHandler(InputElement.KeyDownEvent, d, RoutingStrategies.Tunnel)

File: src/Starter/Views/MainWindow.axaml.cs
using Avalonia;
using Avalonia.Controls;
using Avalonia.Data.Converters;
using Avalonia.Input;
using Avalonia.Interactivity;
using Avalonia.Threading;
using Avalonia.VisualTree;
using System.Globalization;
using Avalonia.Platform;
using R3;
using Serilog;
using Starter.Controls;
using Starter.Features;
using Starter.Features.PlatformInterop;
using Starter.ViewModels;

namespace Starter.Views;

public class FirstNonNullConverter : IMultiValueConverter
{
    public object? Convert(
        IList<object?> values,
        Type targetType,
        object? parameter,
        CultureInfo culture
    ) => values.OfType<object>().FirstOrDefault();
}

public class TimeSpanConverter : IValueConverter
{
    public object Convert(object? value, Type targetType, object? parameter, CultureInfo culture)
    {
        if (value is not TimeSpan t) return string.Empty;

        if ((int)t.TotalMilliseconds > 0)
            return $"{t.TotalMilliseconds.ToString("N2", culture)} ms";

        if ((int)t.TotalMicroseconds > 0)
            return $"{t.TotalMilliseconds.ToString("N2", culture)} μs";

        return $"{t.TotalNanoseconds.ToString("N2", culture)} ns";
    }

    public object? ConvertBack(object? value, Type targetType, object? parameter, CultureInfo culture) => null;
}

public class IsNotZeroConverter : IValueConverter
{
    public object? Convert(object? value, Type targetType, object? parameter, CultureInfo culture) =>
        value is int i
            ? i > 0
            : null;

    public object? ConvertBack(object? value, Type targetType, object? parameter, CultureInfo culture) => null;
}

public partial class MainWindow : TranslucentWindow
{
    private MainWindowViewModel vm = null!;
    private readonly IPlatformInterop platformInterop;
    private readonly IPlatformHandle platformHandle;

    private readonly Stack<(string? text, int caretPosition, object? selectedItem)> previousSelectionState = new();

    private ListBox CurrentResultList =>
        vm.ContextMenuActivated
            ? ContextMenuResultList
            : ResultList;

    public MainWindow()
    {
        // Dummy constructor to prevent XAML warnings
        platformHandle = null!;
        throw new Exception("This constructor should never be called");
    }

    public MainWindow(IPlatformInterop platformInterop)
    {
        this.platformInterop = platformInterop;

        var platformHandle = TryGetPlatformHandle();
        if (platformHandle is null)
        {
            Log.Error("Failed to get platform handle");
            throw new Exception("Failed to get platform handle");
        }
        this.platformHandle = platformHandle;

        InitializeComponent();
        platformInterop.SetupHotkeyCallback(this);
        TextBox.AddHandler(KeyDownEvent, TextBox_OnKeyDown, RoutingStrategies.Tunnel);

        Closing += (_, args) =>
        {
            if (args.IsProgrammatic) return;
            args.Cancel = true;
            Hide();
        };

        Activated += (_, _) => OnActivated();
#if !DEBUG
        Deactivated += (_, _) => Hide();
#endif
    }

    protected override void OnPropertyChanged(AvaloniaPropertyChangedEventArgs change)
    {
        base.OnPropertyChanged(change);
        if (change.Property != IsVisibleProperty) return;
        if (!change.GetNewValue<bool>()) return;
        vm.GreetingVm.RefreshGreetingCommand.Execute(null);
    }

    protected override void OnLoaded(RoutedEventArgs e)
    {
        base.OnLoaded(e);
        if (DataContext is not MainWindowViewModel dataContext) return;
        vm = dataContext;

        SetupKeyboardShortcuts();

        // Subscribe to view model commands
        vm.ClearTextBox += (_, prefixLength) => Dispatcher.UIThread.Post(() =>
        {
            TextBox.CaretIndex = 0; // Change caret index before text prevents a color bug
            TextBox.Text = TextBox.Text?[prefixLength..];
        });
        vm.HideWindow += (_, _) => Dispatcher.UIThread.Post(Hide);

        // Bind zoomed mode
        vm.Config
            .Select(config => config.ZoomedMode)
            .DistinctUntilChanged()
            .Subscribe(SetResourceDictionary);
        SetResourceDictionary(vm.Config.Value.ZoomedMode);

        // Refresh keyboard shortcut when needed
        if (!platformInterop.HotkeyRegistrable) return;
        vm.Config
            .Select(config => config.KeyboardShortcut)
            .DistinctUntilChanged()
            .Subscribe(shortcut => platformInterop.RegisterHotkey(shortcut, this));
        platformInterop.RegisterHotkey(vm.Config.Value.KeyboardShortcut, this);
    }

    private void OnActivated()
    {
        // Center window
        var screen = Screens.ScreenFromTopLevel(this) ?? Screens.Primary;
        if (screen is null)
        {
            Log.Error("No screen available. Can't center window");
            return;
        }

        Position =
            screen.WorkingArea.TopLeft
            + new PixelPoint(
                (int)Math.Round((screen.WorkingArea.Width - Width * screen.Scaling) / 2.0),
                (int)Math.Round((screen.WorkingArea.Height - 470 * screen.Scaling) / 2.0)
            );

        // Reset focus
        TextBox.Focus();
        TextBox.SelectAll();

        // Reset selection
        ResultList.Selection.Select(0);

        // Scroll to top to preserve the top padding
        if (ResultList.ItemCount != 0)
            ResultList.Scroll?.Offset = new Vector(0, 0); // Scroll to home
    }

    private void SetResourceDictionary(bool zoomedMode)
    {
        var found = this.TryFindResource(
            zoomedMode ? "ZoomedMode" : "NormalMode",
            ActualThemeVariant,
            out var resourceDict
        );

        if (!found || resourceDict is not IResourceDictionary v) return;
        Resources = v;
    }

    private void SetupKeyboardShortcuts()
    {
        // Add escape key binding
        var onEscape = new RelayCommand(() =>
        {
            if (vm.ContextMenuActivated) CloseContextMenu();
            else if (vm.Activator is not null) vm.ResetActivatorCommand.Execute(null);
            else Dispatcher.UIThread.Post(Hide);
        });

        KeyBindings.Add(new KeyBinding
        {
            Command = onEscape,
            Gesture = new KeyGesture(Key.Escape)
        });
    }

    private void SelectContextMenuResult(object? result)
    {
        if (result is not ContextMenuResultData data) return;
        if (!data.TryGetEntry(out var entry)) return;

        previousSelectionState.Push((
            TextBox.Text,
            TextBox.CaretIndex,
            CurrentResultList.SelectedItem
        ));

        var openedNewContextMenu = vm.SelectContextMenuResult(entry, platformHandle);
        if (!openedNewContextMenu) previousSelectionState.Pop();
    }

    private void OpenContextMenu(object? result)
    {
        if (result is not SearchResultData data) return;

        previousSelectionState.Push((
            TextBox.Text,
            TextBox.CaretIndex,
            CurrentResultList.SelectedItem
        ));

        var openedNewContextMenu = vm.OpenContextMenu(data);
        if (!openedNewContextMenu) previousSelectionState.Pop();
    }

    private void CloseContextMenu()
    {
        vm.CloseContextMenuCommand.Execute(null);
        if (!previousSelectionState.TryPop(out var state)) return;

        TextBox.Text = state.text;
        TextBox.CaretIndex = 0; // Fixes caret visual position bug
        TextBox.CaretIndex = state.caretPosition;
        CurrentResultList.SelectedItem = state.selectedItem;
    }

    private void ResultList_OnPointerReleased(object? sender, PointerReleasedEventArgs e)
    {
        var pos = e.GetPosition(this);
        var clickedControl =
            this.GetVisualsAt(pos)
                .FirstOrDefault(v => v.DataContext is SearchResultData or ContextMenuResultData);

        if (clickedControl is null) return;

        if (clickedControl.DataContext is SearchResultData)
            vm.SelectResultCommand.Execute(clickedControl.DataContext);

        if (clickedControl.DataContext is ContextMenuResultData { IsSeparator: false })
            SelectContextMenuResult(clickedControl.DataContext);
    }

    private void TextBox_OnKeyDown(object? sender, KeyEventArgs e)
    {
        switch (e.Key)
        {
            case Key.Enter:
                if (vm.ContextMenuActivated)
                    SelectContextMenuResult(ContextMenuResultList.SelectedItem);
                else
                    vm.SelectResultCommand.Execute(ResultList.SelectedItem);

                return;

            // If caret is at start and backspace
            case Key.Back when TextBox.CaretIndex == 0:
                if (vm.ContextMenuActivated) // Close context menu
                {
                    CloseContextMenu();
                    e.Handled = true;
                    return;
                }

                if (vm.Activator is not null) // Disable activator
                {
                    vm.ResetActivatorCommand.Execute(null);
                    e.Handled = true;
                    return;
                }

                break;

            // Tab opens the context menu
            case Key.Tab:
                if (vm.ContextMenuActivated)
                    SelectContextMenuResult(ContextMenuResultList.SelectedItem);
                else
                    OpenContextMenu(ResultList.SelectedItem);

                e.Handled = true;
                return;
        }

        // Set custom keyboard navigation
        // -> The goal is to be able to navigate in the listbox without losing the focus on the textbox
        int newSelectedIdx;

        if (e.Key.ToNavigationDirection() == NavigationDirection.Up)
            newSelectedIdx = Math.Max(CurrentResultList.SelectedIndex - 1, 0);
        else if (e.Key.ToNavigationDirection() == NavigationDirection.Down)
            newSelectedIdx = Math.Min(CurrentResultList.SelectedIndex + 1, CurrentResultList.ItemCount - 1);
        else return;

        // TODO: Always keep bottom padding
        // Scroll to top or bottom to preserve the paddings
        if (CurrentResultList.Scroll is not null)
        {
            if (newSelectedIdx == 0) CurrentResultList.Scroll.Offset = new Vector(0, 0);
            else if (newSelectedIdx == CurrentResultList.ItemCount - 1) CurrentResultList.Scroll.Offset = new Vector(0, CurrentResultList.Scroll.Extent.Height);
        }

        // Select next item
        CurrentResultList.SelectedIndex = newSelectedIdx;

        // Prevent focusing a separator
        if (vm.ContextMenuActivated
            && ContextMenuResultList.SelectedIndex != CurrentResultList.ItemCount - 1
            && ContextMenuResultList.Selection.SelectedItem is ContextMenuResultData { IsSeparator: true })
            TextBox_OnKeyDown(sender, e);

        e.Handled = true;
    }
}
